src/java/main/org/apache/zookeeper/client/StaticHostProvider.java

```diff
@@ -47,59 +51,169 @@
-                    // the toString() method.
-                    if (resolvedAddress.toString().startsWith("/")
-                            && resolvedAddress.getAddress() != null) {
-                        this.serverAddresses.add(
-                                new InetSocketAddress(InetAddress.getByAddress(
-                                        address.getHostName(),
-                                        resolvedAddress.getAddress()),
-                                        address.getPort()));
-                    } else {
-                        this.serverAddresses.add(new InetSocketAddress(resolvedAddress.getHostAddress(), address.getPort()));
-                    }
+                    this.serverAddresses.add(new InetSocketAddress(resolvedAddress, address.getPort()));
                 }
             } catch (UnknownHostException e) {
                 LOG.error("Unable to connect to server: {}", address, e);
             }
         }
-        
+
         if (this.serverAddresses.isEmpty()) {
             throw new IllegalArgumentException(
                     "A HostProvider may not be empty!");
         }
+
         Collections.shuffle(this.serverAddresses);
     }
 
+    /**
+     * Evaluate to a hostname if one is available and otherwise it returns the
+     * string representation of the IP address.
+     *
+     * In Java 7, we have a method getHostString, but earlier versions do not support it.
+     * This method is to provide a replacement for InetSocketAddress.getHostString().
+     *
+     * @param addr
+     * @return Hostname string of address parameter
+     */
+    private String getHostString(InetSocketAddress addr) {
+        String hostString = "";
+
+        if (addr == null) {
+            return hostString;
+        }
+        if (!addr.isUnresolved()) {
+            InetAddress ia = addr.getAddress();
+
+            // If the string starts with '/', then it has no hostname
+            // and we want to avoid the reverse lookup, so we return
+            // the string representation of the address.
+            if (ia.toString().startsWith("/")) {
+                hostString = ia.getHostAddress();
+            } else {
+                hostString = addr.getHostName();
+            }
+        } else {
+            // According to the Java 6 documentation, if the hostname is
+            // unresolved, then the string before the colon is the hostname.
+            String addrString = addr.toString();
+            hostString = addrString.substring(0, addrString.lastIndexOf(':'));
+        }
+
+        return hostString;
+    }
+
     public int size() {
         return serverAddresses.size();
     }
 
+    // Counts the number of addresses added and removed during
+    // the last call to next. Used mainly for test purposes.
+    // See StasticHostProviderTest.
+    private int nextAdded = 0;
```

Comment:
instead of exposing all these metrics, which involve making changes to application logic. Why don't we just expose the `serverAddresses` to the test?